File: python/test/t_TruncatedNormalFactory_std.py
#! /usr/bin/env python

from __future__ import print_function
import openturns as ot
import openturns.testing as ott
from math import sqrt

ot.TESTPREAMBLE()
ot.RandomGenerator.SetSeed(0)


mu = [0.0] * 4
sigma = [1.0] * 4
a = [-4., -1., 1., 3.]
b = [4., 4., 2., 6.]

for i in range(4):
    ot.PlatformInfo.SetNumericalPrecision(1 if i == 2 else 2)
    distribution = ot.TruncatedNormal(mu[i], sigma[i], a[i], b[i])
    size = 10000
    sample = distribution.getSample(size)
    factory = ot.TruncatedNormalFactory()
    estimatedDistribution = factory.build(sample)
    print("distribution=", repr(distribution))
    print("Estimated distribution=", repr(estimatedDistribution))
    estimatedDistribution = factory.build()
    print("Default distribution=", estimatedDistribution)
    estimatedDistribution = factory.build(
        distribution.getParameter())
    print("Distribution from parameters=", estimatedDistribution)
    estimatedTruncatedNormal = factory.buildAsTruncatedNormal(sample)
    print("TruncatedNormal          =", distribution)
    print("Estimated TruncatedNormal=", estimatedTruncatedNormal)
    estimatedTruncatedNormal = factory.buildAsTruncatedNormal()
    print("Default TruncatedNormal=", estimatedTruncatedNormal)
    estimatedTruncatedNormal = factory.buildAsTruncatedNormal(
        distribution.getParameter())
    print("TruncatedNormal from parameters=", estimatedTruncatedNormal)
    # Various estimators
    estimatedDistribution = factory.build(sample)
    print('Estimated distribution with default estimator =', estimatedDistribution)
    estimatedDistribution = factory.buildMethodOfMoments(sample)
    print('Estimated distribution with moments=', estimatedDistribution)
    estimatedDistribution = factory.buildMethodOfLikelihoodMaximization(sample)
    print('Estimated distribution with likelihoodMax.=', estimatedDistribution)
sample = [[0.0]] * size
estimatedDistribution = factory.build(sample)
print("Estimated distribution with default estimator=", repr(estimatedDistribution))
sample = [[1.0]] * size
estimatedDistribution = factory.build(sample)
print("Estimated distribution with default estimator=", repr(estimatedDistribution))

# Build method of moments
print("Build method of moments")
size = 10000
distribution = ot.TruncatedNormal(2.0, 3.0, -1.0, 4.0)
sample = distribution.getSample(size)
factory = ot.TruncatedNormalFactory()
estimatedTN = factory.buildMethodOfMoments(sample)
exact_mu = sample.computeMean()[0]
exact_sigma2 = sample.computeCovariance()[0,0]
exact_a = sample.getMin()[0]
exact_b = sample.getMax()[0]
computed_mu = estimatedTN.getMean()[0]
computed_sigma2 = estimatedTN.getCovariance()[0,0]
computed_a = estimatedTN.getA()
computed_b = estimatedTN.getB()
ott.assert_almost_equal(exact_mu, computed_mu, 1.e-2, 0.0)
ott.assert_almost_equal(exact_sigma2, computed_sigma2, 1.e-3, 0.0)
ott.assert_almost_equal(exact_a, computed_a, 0.0, 10.0 / size)
ott.assert_almost_equal(exact_b, computed_b, 0.0, 10.0 / size)

# Build method of likelihood maximization
print("Build method of likelihood maximization")
size = 10000
distribution = ot.TruncatedNormal(2.0, 3.0, -1.0, 4.0)
sample = distribution.getSample(size)
factory = ot.TruncatedNormalFactory()
estimatedTN = factory.buildMethodOfLikelihoodMaximization(sample)
exact_mu = distribution.getMu()
exact_sigma = distribution.getSigma()
exact_a = distribution.getA()
exact_b = distribution.getB()
computed_mu = estimatedTN.getMu()
computed_sigma = estimatedTN.getSigma()
computed_a = estimatedTN.getA()
computed_b = estimatedTN.getB()
ott.assert_almost_equal(exact_mu, computed_mu, 0.0, 20.0 / sqrt(size))
ott.assert_almost_equal(exact_sigma, computed_sigma, 0.0, 20.0 / sqrt(size))
ott.assert_almost_equal(exact_a, computed_a, 0.0, 20.0 / size)
ott.assert_almost_equal(exact_b, computed_b, 0.0, 20.0 / size)

# A tricky case
print("A tricky case")
data = [0.6852,0.9349,0.5884,1.727,1.581,0.3193,-0.5701,1.623,2.210,
        -0.3440,-0.1646]
sample = ot.Sample(data, 1)
factory = ot.TruncatedNormalFactory()
estimatedTN = factory.buildMethodOfLikelihoodMaximization(sample)
exact_mu = 0.639863
exact_sigma = 1.89489
exact_a = -0.783954
exact_b = 2.42385
computed_mu = estimatedTN.getMu()
computed_sigma = estimatedTN.getSigma()
computed_a = estimatedTN.getA()
computed_b = estimatedTN.getB()
ott.assert_almost_equal(exact_mu, computed_mu, 1.e-2, 0.0)
ott.assert_almost_equal(exact_sigma, computed_sigma, 1.e-2, 0.0)
ott.assert_almost_equal(exact_a, computed_a, 1.e-2, 0.0)
ott.assert_almost_equal(exact_b, computed_b, 1.e-2, 0.0)

File: lib/test/t_TruncatedNormalFactory_std.cxx
//                                               -*- C++ -*-
/**
 *  @brief The test file of class TruncatedNormal for standard methods
 *
 *  Copyright 2005-2022 Airbus-EDF-IMACS-ONERA-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include "openturns/OT.hxx"
#include "openturns/OTtestcode.hxx"

using namespace OT;
using namespace OT::Test;

int main(int, char *[])
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);
  setRandomGenerator();

  try
  {
    Point mu(4, 0.0);
    Point sigma(4, 1.0);
    Point a(4);
    Point b(4);
    a[0] = -4.0;
    b[0] = 4.0;
    a[1] = -1.0;
    b[1] = 4.0;
    a[2] = 1.0;
    b[2] = 2.0;
    a[3] = 3.0;
    b[3] = 6.0;
    for (UnsignedInteger i = 0; i < 4; i++)
    {
      PlatformInfo::SetNumericalPrecision(i == 2 ? 1 : 2);
      TruncatedNormal distribution(mu[i], sigma[i], a[i], b[i]);
      UnsignedInteger size = 10000;
      Sample sample(distribution.getSample(size));
      TruncatedNormalFactory factory;
      CovarianceMatrix covariance;
      // Distribution estimatedDistribution(factory.build(sample, covariance));
      Distribution estimatedDistribution(factory.build(sample));
      fullprint << "Distribution          =" << distribution << std::endl;
      fullprint << "Estimated distribution=" << estimatedDistribution << std::endl;
      // fullprint << "Covariance=" << covariance << std::endl;
      estimatedDistribution = factory.build();
      fullprint << "Default distribution=" << estimatedDistribution << std::endl;
      estimatedDistribution = factory.build(distribution.getParameter());
      fullprint << "Distribution from parameters=" << estimatedDistribution << std::endl;
      TruncatedNormal estimatedTruncatedNormal(factory.buildAsTruncatedNormal(sample));
      fullprint << "TruncatedNormal          =" << distribution << std::endl;
      fullprint << "Estimated TruncatedNormal=" << estimatedTruncatedNormal << std::endl;
      estimatedTruncatedNormal = factory.buildAsTruncatedNormal();
      fullprint << "Default TruncatedNormal=" << estimatedTruncatedNormal << std::endl;
      estimatedTruncatedNormal = factory.buildAsTruncatedNormal(distribution.getParameter());
      fullprint << "TruncatedNormal from parameters=" << estimatedTruncatedNormal << std::endl;
    }
    // Test for constant sample
    TruncatedNormalFactory factory;
    UnsignedInteger size = 10000;
    Sample sample(size, Point(1, 0.0));
    // buildMethodOfMoments
    fullprint << "buildMethodOfMoments" << std::endl;
    TruncatedNormal distribution(2.0, 3.0, -1.0, 4.0);
    Sample sample2(distribution.getSample(size));
    TruncatedNormal estimatedTN(factory.buildMethodOfMoments(sample2));
    fullprint << "Estimated from moments=" << estimatedTN << std::endl;
    const Scalar sample_mu = sample2.computeMean()[0];
    const Scalar sample_sigma2 = sample2.computeCovariance()(0, 0);
    const Scalar sample_a = sample2.getMin()[0];
    const Scalar sample_b = sample2.getMax()[0];
    Scalar computed_mu = estimatedTN.getMean()[0];
    Scalar computed_sigma2 = estimatedTN.getCovariance()(0, 0);
    Scalar computed_a = estimatedTN.getA();
    Scalar computed_b = estimatedTN.getB();
    assert_almost_equal(sample_mu, computed_mu, 1.e-2, 0.0);
    assert_almost_equal(sample_sigma2, computed_sigma2, 1.e-3, 0.0);
    assert_almost_equal(sample_a, computed_a, 0.0, 10.0 / size);
    assert_almost_equal(sample_b, computed_b, 0.0, 10.0 / size);
    // buildMethodOfLikelihoodMaximization
    fullprint << "buildMethodOfLikelihoodMaximization" << std::endl;
    TruncatedNormal distribution3(2.0, 3.0, -1.0, 4.0);
    Sample sample3(distribution3.getSample(size));
    TruncatedNormal estimatedTN3(factory.buildMethodOfLikelihoodMaximization(sample3));
    fullprint << "Estimated from likelihoodMaximization=" << estimatedTN3 << std::endl;
    const Scalar exact_mu = distribution3.getMu();
    const Scalar exact_sigma = distribution3.getSigma();
    const Scalar exact_a = distribution3.getA();
    const Scalar exact_b = distribution3.getB();
    computed_mu = estimatedTN3.getMu();
    const Scalar computed_sigma = estimatedTN3.getSigma();
    computed_a = estimatedTN3.getA();
    computed_b = estimatedTN3.getB();
    assert_almost_equal(computed_mu, exact_mu, 0.0, 20.0 / sqrt(size));
    assert_almost_equal(computed_sigma, exact_sigma, 0.0, 20.0 / sqrt(size));
    assert_almost_equal(computed_a, exact_a, 0.0, 20.0 / size);
    assert_almost_equal(computed_b, exact_b, 0.0, 20.0 / size);
  }
  catch (TestFailed & ex)
  {
    std::cerr << ex << std::endl;
    return ExitCode::Error;
  }


  return ExitCode::Success;
}
